### PR TITLE
bug: LT-1443 - Refactor Notifications

### PR DESCRIPTION
**Pull request by andytaylorkainos**

### CODE REVIEW

File: conf/constants.py [hunk 1]
```diff
@@ -50,10 +50,10 @@
 OPEN_LICENCE = "open_licence"
 HMRC_QUERY = "hmrc_query"
 
-# Notification constants
-CASE_NOTE = "case_note"
-ECJU_QUERY = "ecju_query"
-GENERATED_CASE_DOCUMENT = "generated_case_document"
+# Case type constants
+APPLICATION = "application"
```
Comment (jfaracik) [inline conf/constants.py]: Make this a class (or enum) like Permission underneath

Comment (andytaylorkainos) [reply conf/constants.py]: Done

File: core/services.py [hunk 2]
```diff
@@ -83,12 +83,13 @@ def post_external_locations(request, pk, json):
...
 
-def get_notifications(request, unviewed):
-    url = NOTIFICATIONS_URL
-    if unviewed:
-        url = "%s?unviewed=True" % url
+def get_notifications(request, content_types=None, count_only=False):
+    url = f"{NOTIFICATIONS_URL}?count_only={count_only}"
+    if content_types:
+        content_types = ",".join(content_types)
+        url = f"{url}&content_types={content_types}"
```
Comment (jfaracik) [inline core/services.py]: Convert to query params should be able to handle this for you, as opposed to combining it yourself.

Comment (andytaylorkainos) [reply core/services.py]: Done

File: templates/goods/good.html [hunk 3]
```diff
@@ -41,15 +41,15 @@ <h2 class="govuk-heading-l">
 				<br>
 			{% endif %}
 
-			{% if good.query_id %}
+			{% if good.query.id %}
```
Comment (Mark-Spiers-kainos) [inline templates/goods/good.html]: could you not just do if `good.query?`

Comment (andytaylorkainos) [reply templates/goods/good.html]: Done

File: core/helpers.py [hunk 4]
```diff
@@ -35,10 +34,10 @@ def str_date_only(value):
...
-def generate_notification_string(notifications, object_type):
-    notifications_count = len(
-        [x for x in notifications if x["object_type"] == object_type or x["parent_type"] == object_type]
-    )
+def generate_notification_string(notifications, case_types):
+    notifications_count = 0
+    for case_type in case_types:
+        notifications_count += notifications["notifications_count"].get(case_type, 0)
 
     if notifications_count == 0:
```
Comment (stephanieboyd) [inline core/helpers.py]: `if not notifications_count`

File: conf/constants.py [hunk 5]
```diff
@@ -76,3 +71,9 @@ class Permissions:
     SUBMIT_LICENCE_APPLICATION = "SUBMIT_LICENCE_APPLICATION"
 
     MANAGE_ORGANISATION_PERMISSIONS = [ADMINISTER_SITES, ADMINISTER_USERS, EXPORTER_ADMINISTER_ROLES]
+
+
+class CaseType:
+    APPLICATION = "application"
```
Comment (stephanieboyd) [inline conf/constants.py]: Should HMRC be in this list?

File: core/helpers.py [hunk 6]
```diff
@@ -35,10 +34,10 @@ def str_date_only(value):
     return return_value.strftime("%d %B %Y")
 
 
-def generate_notification_string(notifications, object_type):
-    notifications_count = len(
-        [x for x in notifications if x["object_type"] == object_type or x["parent_type"] == object_type]
-    )
+def generate_notification_string(notifications, case_types):
+    notifications_count = 0
```
Comment (stephanieboyd) [inline core/helpers.py]: Why not pull out `notifications["notifications_count"]`, do that once. If that's a dict - you could also then (maybe, from where I'm sitting):

```
>>> j = {"l": 1, "k":1}
>>> types = ["l", "k"]
>>> sum([value for case, value in j.items() if case in types])```

File: core/services.py [hunk 7]
```diff
@@ -83,12 +83,14 @@ def post_external_locations(request, pk, json):
     return data.json(), data.status_code
 
 
-def get_notifications(request, unviewed):
-    url = NOTIFICATIONS_URL
-    if unviewed:
-        url = "%s?unviewed=True" % url
+def get_notifications(request, case_types=None, count_only=True):
+    url = f"{NOTIFICATIONS_URL}?count_only={count_only}"
```
Comment (stephanieboyd) [inline core/services.py]: What's count only?

Comment (stephanieboyd) [reply core/services.py]: Maybe give this function a docstring to explain, it's not immediately obvious to me what/why

File: core/services.py [hunk 8]
```diff
@@ -83,12 +83,14 @@ def post_external_locations(request, pk, json):
...
+
+    if case_types:
+        url = f"{url}&{convert_value_to_query_param(key='case_type', value=case_types)}"
+
     data = get(request, url)
-    return data.json().get("results")
+    return data.json()
```
Comment (stephanieboyd) [inline core/services.py]: No status code?

File: core/views.py [hunk 9]
```diff
@@ -64,17 +64,17 @@ def get(self, request, **kwargs):
                     [
                         Tile(
                             get_string("applications.title"),
-                            generate_notification_string(notifications, "base_application"),
+                            generate_notification_string(notifications, case_types=[CaseType.APPLICATION]),
                             reverse_lazy("applications:applications"),
                         ),
                         Tile(
                             "Goods",
```
Comment (stephanieboyd) [inline core/views.py]: Should these titles be in strings?

File: templates/applications/application.html [hunk 10]
```diff
@@ -72,26 +72,26 @@ <h1 class="govuk-heading-l">{{ title }}</h1>
 			{% if application.application_type.key != 'hmrc_query' %}
 				<a href="{% url 'applications:application' application.id 'case-notes' %}" class="lite-tabs__tab {% if type == 'case-notes' %}lite-tabs__tab--selected{% endif %}" id="link-case-notes">
 					Notes
-					{% if case_note_notifications %}
+					{% if application.exporter_user_notifications_count.casenote > 0 %}
```
Comment (stephanieboyd) [inline templates/applications/application.html]: I'd drop the `> 0`

File: templates/goods/goods.html [hunk 11]
```diff
@@ -45,11 +45,11 @@ <h2 class="govuk-heading-l">
 					<tr class="govuk-table__row">
                         <td class="govuk-table__cell">
 							<a class="govuk-link govuk-link--no-visited-state app-icon-label" href="{% url 'goods:good' pk=good.id %}">
-								<p>{{ good.description }}</p>
-								{% if good.status.key == 'verified' %}<span class="app-verified__symbol">{% svg 'verified' %}</span>{% endif %}
-								{% if good.query_id and good.query_id in notifications %}
-									<span class="lite-notification-bubble">{{ notifications|key_value:good.query_id }} <span class="govuk-visually-hidden"> {% lcs "NOTIFICATIONS" %}</span></span>
+								{{ good.description }}
+								{% if good.query.exporter_user_notifications_count.total > 0 %}
+									<span class="lite-notification-bubble">{{ good.query.exporter_user_notifications_count.total }} <span class="govuk-visually-hidden"> {% lcs "NOTIFICATIONS" %}</span></span>
```
Comment (stephanieboyd) [inline templates/goods/goods.html]: Are we moving to class based strings now?

File: conf/constants.py [hunk 12]
```diff
@@ -76,3 +71,9 @@ class Permissions:
...
+    CLC_QUERY = "clc_query"
+    EUA_QUERY = "end_user_advisory_query"
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply conf/constants.py]: Isnt needed yet

File: conf/constants.py (hunk 5 above)
Comment (andytaylorkainos) [reply conf/constants.py]: Isnt used yet

File: core/services.py (hunk 7 above)
Comment (andytaylorkainos) [reply core/services.py]: an API query param to return only the number of notifications, not the notification data

File: core/views.py (hunk 9 above)
Comment (andytaylorkainos) [reply core/views.py]: Didnt touch them :)

File: templates/goods/goods.html (hunk 11 above)
Comment (andytaylorkainos) [reply templates/goods/goods.html]: Daaamn i hit one

File: templates/applications/application.html (hunk 10 above)
Comment (andytaylorkainos) [reply templates/applications/application.html]: Done

File: core/services.py (hunk 8 above)
Comment (andytaylorkainos) [reply core/services.py]: Done

File: core/services.py (hunk 7 above)
Comment (andytaylorkainos) [reply core/services.py]: dosctring added

File: conf/constants.py (hunk 5 above)
Comment (andytaylorkainos) [reply conf/constants.py]: Actually - HMRC is an application type (annoying i know)

File: conf/constants.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply conf/constants.py]: Actually - HMRC is an application type (annoying i know)

File: core/helpers.py (hunk 4 above)
Comment (andytaylorkainos) [reply core/helpers.py]: done

File: core/helpers.py (hunk 6 above)
Comment (andytaylorkainos) [reply core/helpers.py]: Done

File: templates/goods/goods.html (hunk 11 above)
Comment (andytaylorkainos) [reply templates/goods/goods.html]: Done

File: core/views.py (hunk 9 above)
Comment (stephanieboyd) [reply core/views.py]: We said files, my good sir

Comment (stephanieboyd) [reply core/views.py]: (sorry)

Comment (andytaylorkainos) [reply core/views.py]: nooooo

File: conf/constants.py (hunk 5 above)
Comment (stephanieboyd) [reply conf/constants.py]: Naw, it is it's own type - 

```"application"	"Application"
"clc_query"	"CLC Query"
"end_user_advisory_query"	"End User Advisory Query"
"hmrc_query"	"HMRC Query"```

Comment (stephanieboyd) [reply conf/constants.py]: I'd probably still push for adding it here, even if not used - but okay

File: core/views.py (hunk 9 above)
Comment (andytaylorkainos) [reply core/views.py]: done

File: conf/constants.py (hunk 5 above)
Comment (andytaylorkainos) [reply conf/constants.py]: done

**Pull request merged**